Frontend/src/Components/Dashboard/GraphContainer.js

```diff
@@ -0,0 +1,253 @@
+      // if (graphTitles[idx] && graphTitles[idx] !== "Custom")
+      //   onSave(graphTitles[idx], datasets);
+
+      // duplicate the graph titles
+      const copy = graphTitles.slice();
+
+      // check for special values
+      if (name?.length && name !== "Custom") {
+        const oldIdx = graphTitles.indexOf(name);
+
+        // found old position of graph, swap with new position
+        if (oldIdx !== -1) {
+          // see https://jsbench.me/snl20xx1g9/1 for reasoning behind this algorithm
+          copy[oldIdx] = copy[idx];
+          copy[idx] = name;
+          // console.log(graphTitles, "[", idx, "] =", name, "=", graphTitles);
+
+          setGraphTitles(copy);
+          return;
+        }
+      }
+
+      // default set: if new custom/empty graph or replacing new custom/empty graph
+      // see https://jsbench.me/vsl20xqpso/1 for reasoning behind this algorithm
+      copy[idx] = name;
+      // console.log(graphData, "[", idx, "] =", name, "=", graphData);
+      setGraphTitles(copy);
+    },
+    [graphTitles, setGraphTitles]
+  );
+
+  //------------------------- Saving custom graphs ----------------------------
+  const [customGraphData, setCustomGraphData] = useState({});
+  // console.log("custom graph data:", customGraphData);
+
+  /**
+   * Saves the given graph's data (namely, the datasets that it is displaying)
+   *
+   * @param {string} title the title of the graph to be saved
+   * @param {string[]} datasets collection of datasets that is associated with this graph
+   * @param {number} index the index of the graph that's being saved
+   */
+  const onSave = useCallback(
+    (title, isNew, datasets, index) => {
+      // check duplicates if new
+      if (isNew && Object.keys(customGraphData).includes(title)) {
+        console.warn("No\n", title, "is in", customGraphData);
+        return false;
+      }
+
+      // set new graph data assigned to this graph
+      const tempData = Object.assign({}, customGraphData);
+      tempData[title] = datasets;
+      setCustomGraphData(tempData);
+
+      // update the graph title
+      const tempTitles = graphTitles.slice();
+      tempTitles[index] = title;
+      setGraphTitles(tempTitles);
+
+      // go ahead with saving
+      return true;
+    },
+    [setCustomGraphData, setGraphTitles, customGraphData, graphTitles]
+  );
+
+  //----------------- Calculate all data that can be shown --------------------
+  const packedData = useMemo(() => {
+    // console.log("memo invoked");
+    console.time("unpack graph data");
+    const temp = allDatasets.map((value) => {
+      const output = {
+        key: value.key,
+        label: value.name,
+        data: queue[value.key],
+        borderColor: value.color,
+        backgroundColor: value.color + "b3",
+      };
+      // console.log("packing:", output);
+      return output;
+    });
+    console.timeEnd("unpack graph data");
+    return temp;
+  }, [allDatasets, queue]);
```

Comment:
A bit more detail on packedData (a header comment including the contents/a description of packedData for sure)

Reply:
Something like you did with generateCategories would be awesome

Reply:
Done